File: year4/pyFEM/geometry/ad_triangulator.py
from collections import defaultdict
import math

from pyFEM.geometry import Point, Edge, Polygon
from pyFEM import symbols as sym


def uniq(sorted_seq):
    res = [sorted_seq[0]]
    for i in range(1, len(sorted_seq)):
        if sorted_seq[i] != sorted_seq[i - 1]:
            res.append(sorted_seq[i])
    return res

class MonotonePolygonTriangulator(object):
    """Class for triangulating monotone polygons (with no intruding vertices)"""
    def __init__(self, diameter, polygon):
        self.diameter = diameter
        self.polygon = polygon
        self.triangles = []
        
        self.precalculated_polygon = None
        self.points_to_triangles = None
        self.points_to_points = None
        self.bound_points = None
        self.bounds_hasn = None

    def triangulate(self):
        self.triangles = []
        self.curr_polygon = self.precalculate_polygon()
        self.curr_angles = self.precalculate_angles(self.curr_polygon)
        self.points_to_points = defaultdict(list)
        self.points_to_triangles = defaultdict(list)
        
        self.inner_triangulate()
        
        return self.triangles

    def precalculate_polygon(self):
        diameter = self.diameter
        self.bounds_hash = {}
        my_polygon = Polygon()
        for i in range(len(self.polygon)):
            point1 = self.polygon.points[i - 1]
            point2 = self.polygon.points[i]
            my_polygon.extend(Edge.split(point1, point2, diameter))
        self.precalculated_polygon = Polygon.copy(my_polygon.points)
        return my_polygon

    def get_bound_points(self, subdivide_n=0):
        self.bound_points = {}

        precalculated_copy = Polygon.copy(self.precalculated_polygon.points)
        precalculated_copy.subdivide(subdivide_n)
        
        for p in precalculated_copy:
            self.bound_points[p] = True
        return self.bound_points

    def precalculate_angles(self, polygon):
        angles = {}
        for i in range(len(polygon)):
            angles[polygon.points[i]] = polygon.get_angle(i)
        return sorted(angles.items(), cmp=lambda a, b: cmp(a[1], b[1]))

    def inner_find_point(self, p1, p2, p3, s2, s3, s, angle):
        aa = s3*s3 + s*s - 2*s3*s*math.cos(angle/2.0)
        bb = s2*s2 + s*s - 2*s2*s*math.cos(angle/2.0)

        k = s*s - aa + (p2.x**2 - p1.x**2) + (p2.y**2 - p1.y**2)
        l = aa - bb + (p3.x**2 - p2.x**2) + (p3.y**2 - p2.y**2)

        m = 2.0*(p2.x - p1.x)
        n = 2.0*(p2.y - p1.y)
        p = 2.0*(p3.x - p2.x)
        q = 2.0*(p3.y - p2.y)

        det = n*p - q*m
        x_new = (l*n - k*q)/det
        y_new = (k*p - l*m)/det

        return Point(x_new, y_new)

    def process_polygon4(self, polygon):
        angles = [polygon.get_angle(i) for i in range(len(polygon.points))]
        tr = None
        if angles[0] + angles[2] >= angles[1] + angles[3]:
            tr = polygon.split_indices(0, 2)
        else:
            tr = polygon.split_indices(1, 3)

        self.triangles.append(tr)
        self.save_triangle_points(tr, len(self.triangles) - 1)

        self.triangles.append(polygon) 
        self.save_triangle_points(polygon, len(self.triangles) - 1)

        return polygon

    def find_index(self, angle, point, values):
        left, right = 0, len(values) - 1
        while left <= right:
            m = (left + right) // 2
            if values[m][1] == angle: break
            if angle > values[m][1]:
                left = m + 1
            else:
                right = m - 1
        if left <= right:
            return [True, m]
        else:
            return [False, left]

    def find_by_angle(self, angle, point, values):
        index = self.find_index(angle, point, values)
        while (values[index[1] - 1][1] == angle) and (index[1] > 0):
            index[1] -= 1
        return index

    def find_by_point(self, angle, point, values):
        try:
            index = [x[0] for x in values].index(point)
        except ValueError:
            return [False, None]
        return [True, index]

    def delete_vertex(self, angle, point, values):
        index = self.find_by_point(angle, point, values)
        if not index[0]:
            raise ValueError("Vertex to delete not found")
        del values[index[1]]

    def change_alpha(self, point, alpha, new_alpha, values):
        index = self.find_by_point(alpha, point, values)
        if not index[0]:
            raise ValueError("Alpha to change not found")
        if values[index[1]][1] != alpha:
            raise ValueError("Found bad value")

        del values[index[1]]
        new_index = self.find_by_angle(new_alpha, point, values)
        values.insert(new_index[1], [point, new_alpha])
        return True

    def insert_alpha_point(self, alpha, point, values):
        new_index = self.find_by_angle(alpha, point, values)
        values.insert(new_index[1], [point, alpha])

    def cut_off_vertex(self, index, alpha):
        new_angle = Edge.angle_points_orient(self.curr_polygon.next(),
                                             self.curr_polygon.prev(),
                                             self.curr_polygon.prev(2))
        alpha1 = self.curr_polygon.get_angle(index - 1)
        self.change_alpha(self.curr_polygon.prev(), alpha1, new_angle, self.curr_angles)
        new_angle = Edge.angle_points_orient(self.curr_polygon.next(2),
                                             self.curr_polygon.next(),
                                             self.curr_polygon.prev())
        alpha2 = self.curr_polygon.get_angle(index + 1)
        self.change_alpha(self.curr_polygon.next(), alpha2, new_angle, self.curr_angles)
        new_polygon = self.curr_polygon.split_indices(index - 1, index + 1)
        if len(self.curr_polygon) == 3:
            self.curr_polygon, new_polygon = new_polygon, self.curr_polygon

        self.triangles.append(new_polygon)
        self.save_triangle_points(new_polygon, len(self.triangles) - 1)

    def save_triangle_points(self, triangle, index):
        for p in triangle.points:
            self.points_to_triangles[p].append(index)

        for p in triangle.points:
            for inn_p in triangle.points:
                if inn_p == p: continue
                self.points_to_points[p].append(inn_p)
            self.points_to_points[p].sort()
            self.points_to_points[p] = uniq(self.points_to_points[p])

    def process_case2(self, angle):
        p1 = self.curr_polygon.curr_point
        p2 = self.curr_polygon.prev()
        p3 = self.curr_polygon.next()

        s2 = (p3 - p1).length()
        s3 = (p2 - p1).length()
        s1 = (self.curr_polygon.next(2) - self.curr_polygon.next()).length()
        s4 = (self.curr_polygon.prev() - self.curr_polygon.prev(2)).length()

        s = (s1 + (2 * s2) + (2 * s3) + s4) / 6.0
        p = self.inner_find_point(p1, p2, p3, s2, s3, s, angle)
        if self.curr_polygon.has_point(p):
            return p
        raise RuntimeError("doh")

    def process_case3(self, angle):
        l1 = Edge.length(self.curr_polygon.next(), self.curr_polygon.curr_point)
        l2 = Edge.length(self.curr_polygon.curr_point, self.curr_polygon.prev())

        p2 = self.curr_polygon.curr_point
        p1 = self.curr_polygon.prev()
        a = l2

        index = self.curr_polygon.curr_index

        if l1 < l2:
            p1 = self.curr_polygon.curr_point
            p2 = self.curr_polygon.next()
            a = l1
            index += 1

            p_middle = Point.new((p1.x + p2.x)/2.0, (p1.y + p2.y)/2.0)
    
        m = 2.0*(p2.x - p1.x)
        n = 2.0*(p2.y - p1.y)
        k = ((p2.x**2) + (p2.y**2)) - ((p1.x**2) + (p1.y**2))

        p = p2.y - p1.y
        q = p1.x - p2.x
        l = p1.x*p2.y - p2.x*p1.y + Math.sqrt(3.0)*a*a/2.0

        det = n*p - q*m
        x_new = (l*n - k*q)/det
        y_new = (k*p - l*m)/det

        p = Point.new(x_new, y_new)

        if self.curr_polygon.has_point(p):
            return [p, index]

        # find symmetric point
        p_new = Point(2.0*p.x - p_middle.x, 2.0*p.y - p_middle.y)
        if p_new in self.curr_polygon:
            return [p_new, index] 

        raise RuntimeError("doh doh doh 3")

    def inner_triangulate(self):
        count = 0

        if len(self.curr_polygon) < 3:
            return self.curr_polygon 

        if len(self.curr_polygon) == 3:
            self.triangles.append(self.curr_polygon)
            return self.curr_polygon

        if len(self.curr_polygon) == 4:
            return self.process_polygon4(self.curr_polygon)

# main loop
        while self.curr_angles:
            # first stupid/simple situations
            if len(self.curr_polygon) == 4:
                self.process_polygon4(self.curr_polygon)
                break

            if len(self.curr_polygon) == 3:
                self.triangles.append(self.curr_polygon)
                self.save_triangle_points(self.curr_polygon, len(self.triangles) - 1)
                break

            if len(self.curr_polygon) < 3:
                break 

            # and now sophisticated stuff

            # at first delete element with smalest angle
            element = self.curr_angles[0]
            point = element[0]
            angle = element[1]

            del self.curr_angles[0]
            self.curr_polygon.curr_point = point

            # if it is just a cutoff - do it
            # and go to next loop iteration
            if angle < math.pi/2.0:
                #p "upper cutoff"
                self.cut_off_vertex(self.curr_polygon.curr_index, angle)
                continue

            a1 = (math.pi / 2.0)
            a2 = (5.0*math.pi / 6.0)
            a3 = 2.0*math.pi

            bad_point = False
            used_elements = []

            while True: # while bad_point
                if 0 <= angle <= a1:
                     #p "inner cutoff"
                    bad_point = False
                    self.cut_off_vertex(self.curr_polygon.curr_index, angle)
                elif a1 <= angle <= a2:
                    #p 'case 2'
                    p = self.process_case2(angle)
                    if not p:
                        used_elements.append([Point(point.x, point.y), angle])
                        bad_point = True
                    bad_point = False
                    # if all is ok - than we will insert
                    # a new point - need to add first elements
                    if used_elements:
                        self.curr_angles = used_elements + self.curr_angles
                        used_elements = []

                    p1 = self.curr_polygon.curr_point
                    p2 = self.curr_polygon.prev()
                    p3 = self.curr_polygon.next()

                    tr1 = Polygon.copy([p1, p3, p])
                    tr2 = Polygon.copy([p1, p, p2])

                    self.triangles.append(tr1)
                    self.save_triangle_points(tr1, len(self.triangles) - 1)

                    self.triangles.append(tr2)
                    self.save_triangle_points(tr2, len(self.triangles) - 1)

                    angle1 = self.curr_polygon.get_angle(self.curr_polygon.curr_index - 1)
                    angle2 = self.curr_polygon.get_angle(self.curr_polygon.curr_index + 1)

                    self.curr_polygon.replace_current(p)

                    # change angle of previous point
                    new_angle = self.curr_polygon.get_angle(self.curr_polygon.curr_index - 1)
                    self.change_alpha(self.curr_polygon.prev(), angle1, new_angle, self.curr_angles)

                    # add angle of next point
                    new_angle = self.curr_polygon.get_angle(self.curr_polygon.curr_index + 1)
                    self.change_alpha(self.curr_polygon.next(), angle2, new_angle, self.curr_angles)

                    # add angle of current point
                    new_angle = self.curr_polygon.get_angle(self.curr_polygon.curr_index)
                    self.insert_alpha_point(new_angle, self.curr_polygon.curr_point, self.curr_angles)
                elif a2 <= angle <= a3:
                    return
                    # print('case 3')
                    # arr = self.process_case3(angle)
                    # if not arr:
                    #     used_elements.append([Point(point.x, point.y), angle])
                    #     bad_point = True
                    # else:
                    #     bad_point = False            

                    # if used_elements:
                    #     self.curr_angles = used_elements + self.curr_angles
                    #     used_elements = []

                    # p = arr[0]
                    # index = arr[1]

                    # self.curr_polygon.curr_index = index
                    # p2 = self.curr_polygon.curr_point
                    # p1 = self.curr_polygon.prev

                    # tr = Polygon.copy([p1, p2, p])
                    # self.triangles.append(tr)
                    # self.save_triangle_points(tr, len(self.triangles) - 1)

                    # self.curr_polygon.insert_point(index, p)

                    # # insert angle of current vertex (a new vertex)
                    # new_angle = self.curr_polygon.get_angle(index)
                    # self.insert_alpha_point(new_angle, p, self.curr_angles)

                    # new_angle = self.curr_polygon.get_angle(index - 1)
                    # self.change_alpha(self.curr_polygon.prev, angle, new_angle, self.curr_angles)

                    # new_angle = self.curr_polygon.get_angle(index - 2)
                    # self.change_alpha(self.curr_polygon.prevprev, angle, new_angle, self.curr_angles)


                if len(self.curr_angles) <= 4:
                    break

                # break if self.curr_angles.size <= 4

                        #p self.curr_angles
                        #p self.curr_polygon

                if bad_point:
                    element = self.curr_angles[0]
                    point = element[0]
                    angle = element[1]

                    del self.curr_angles[0]

                    self.curr_polygon.curr_point = point

                    print('move next')
                else:
                    if used_elements:
                        self.curr_angles = used_elements + self.curr_angles
                        used_elements.clear

                if not bad_point:
                    break

File: year4/rFEM/geometry/ad_triangulator.rb
require File.expand_path(File.dirname(__FILE__)) + '/point.rb'
require File.expand_path(File.dirname(__FILE__)) + '/edge.rb'
require File.expand_path(File.dirname(__FILE__)) + '/polygon.rb'

# class for triangulating monotone polygons
# (with no intruding vertices)
class MonotonePolygonTriangulator

  attr :diameter, true
  attr :polygon, true
  attr_reader :precalculated_polygon
  attr_reader :triangles
  attr_reader :points_to_triangles
  attr_reader :points_to_points
  attr_reader :bound_points
  attr_reader :bounds_hash

  def initialize(diameter, polygon)
    @diameter = diameter
    @polygon = polygon
    @triangles = []
  end

  def triangulate
    @triangles = []

    @curr_polygon = precalculate_polygon
    # save local copy for smoothing and
    # bounds determining
    
    @curr_angles = precalculate_angles @curr_polygon    

    @points_to_points = Hash.new {|hash,key| hash[key] = []}
    @points_to_triangles = Hash.new {|hash,key| hash[key] = []}

    inner_triangulate

    @triangles
  end

  # pastes a lot of points on
  # polygon sides
  def precalculate_polygon
    diameter = @diameter

    @bounds_hash = {}
    
    my_polygon = Polygon.new
    
    @polygon.size.times do |i|
      point1 = polygon.points[i - 1]
      point2 = polygon.points[i]
      
      my_polygon.concat Edge.split(point1, point2, diameter)
      
    end

    @precalculated_polygon = Polygon.copy(my_polygon.points)
    
    my_polygon
  end

  def get_bound_points
    @bound_points = {}
    @precalculated_polygon.points.each do |p|
      @bound_points[p] = true
    end
    @bound_points
  end

  def provide_smoothing
    @bound_points = {}
    @precalculated_polygon.points.each do |p|
      @bound_points[p] = true
    end

    @smoothed_triangles = []
    @triangles.each do |tr|
      @smoothed_triangles << Polygon.copy(tr.points)
    end

    @points_to_points.each do |key, value|
      # don't move boundary points
      next if @bound_points.has_key? key

      # calculate new point coordinates
      new_point = Polygon.get_center value

      # replace new value in all triangles
      @points_to_triangles[key].each do |i|
        @smoothed_triangles[i].replace_point!(key, new_point)
      end
    end

    @smoothed_triangles
  end

  # private zone here
  
  private

  def precalculate_angles(polygon)
    angles = {}
    #index_copy = polygon.curr_index
    
    polygon.size.times do |i|
      angles[ polygon.points[i] ] = polygon.get_angle(i)
    end

    angles.sort {|a, b| a[1] <=> b[1]}
  end

  def inner_find_point(p1, p2, p3, s2, s3, s, angle)
    aa = s3*s3 + s*s - 2*s3*s*Math.cos(angle/2.0)
    bb = s2*s2 + s*s - 2*s2*s*Math.cos(angle/2.0)
    
    k = s*s - aa + (p2.x**2 - p1.x**2) + (p2.y**2 - p1.y**2)
    l = aa - bb + (p3.x**2 - p2.x**2) + (p3.y**2 - p2.y**2)
    
    m = 2.0*(p2.x - p1.x)
    n = 2.0*(p2.y - p1.y)
    p = 2.0*(p3.x - p2.x)
    q = 2.0*(p3.y - p2.y)
    
    det = n*p - q*m
    x_new = (l*n - k*q)/det
    y_new = (k*p - l*m)/det
    
    Point.new(x_new, y_new)
  end

  def process_polygon4(polygon)
    angles = []
    polygon.points.each_index {|i| angles << polygon.get_angle(i)}
    
    tr = nil
    if (angles[0] + angles[2]) >= (angles[1] + angles[3])
      tr = polygon.split_indices!(0, 2)
    else
      tr = polygon.split_indices!(1, 3)
    end

    @triangles << tr
    save_triangle_points tr, @triangles.size - 1
    
    @triangles << polygon
    save_triangle_points polygon, @triangles.size - 1
    
    polygon
  end

  def find_index(angle, point, values)
    left, right = 0, values.size-1
    while left <= right
      m = (left + right)/2
      #break if values[m][0] == point
      break if values[m][1] == angle
      if angle > values[m][1]
        left = m + 1
      else
        right = m - 1
      end
    end

    if left <= right
      [true, m]
    else
      [false, left]
    end
  end

  def find_by_angle(angle, point, values)
    index = find_index(angle, point, values)
    index[1] -= 1 while (values[index[1] - 1][1] == angle and index[1] > 0)
    index
  end

  def find_by_point(angle, point, values)
    #index = find_index angle, point, values
    index = values.index {|x| x[0] == point} 
    [index.nil? == false, index]
  end

  def delete_vertex(angle, point, values)
    index = find_by_point angle, point, values
    raise 'delete vertex not found' if index[0] == false
    values.delete_at index[1]
  end

  def change_alpha(point, alpha, new_alpha, values)
    index = find_by_point alpha, point, values
    raise 'change alpha not found' if index[0] == false
    raise 'found bad value' unless values[index[1]][1] == alpha

    # delete old point
    values.delete_at index[1]

    # add new point
    new_index = find_by_angle new_alpha, point, values
    values.insert new_index[1], [point, new_alpha]
    true
  end

  def insert_alpha_point(alpha, point, values)
    new_index = find_by_angle alpha, point, values
    values.insert new_index[1], [point, alpha]
  end

  def cut_off_vertex(index, alpha)
    new_angle = Edge.angle_points_orient(@curr_polygon.next, @curr_polygon.prev, @curr_polygon.prevprev)
    alpha1 = @curr_polygon.get_angle(index - 1)
    change_alpha(@curr_polygon.prev, alpha1, new_angle, @curr_angles)
    
    new_angle = Edge.angle_points_orient(@curr_polygon.nextnext, @curr_polygon.next, @curr_polygon.prev)
    alpha2 = @curr_polygon.get_angle(index + 1)
    change_alpha(@curr_polygon.next, alpha2, new_angle, @curr_angles)

    new_polygon = @curr_polygon.split_indices!(index - 1, index + 1)

    if @curr_polygon.size == 3
      @curr_polygon, new_polygon = new_polygon, @curr_polygon
    end

    @triangles << new_polygon
    save_triangle_points(new_polygon, @triangles.size - 1)
  end

  def save_triangle_points(triangle, index)
    triangle.points.each do |p|
      @points_to_triangles[p] << index
      # @points_to_triangles[p].sort!.uniq!
    end
    
    triangle.points.each do |p|
      triangle.points.each do |inn_p|
        next if inn_p == p
        
        @points_to_points[p] << inn_p
      end
      @points_to_points[p].sort!.uniq!
    end
  end

  
  def process_case2(angle)

    p1 = @curr_polygon.curr_point
    p2 = @curr_polygon.prev
    p3 = @curr_polygon.next

    s2 = Edge.length(p3, p1)
    s3 = Edge.length(p2, p1)
    s1 = Edge.length(@curr_polygon.nextnext, @curr_polygon.next)
    s4 = Edge.length(@curr_polygon.prev, @curr_polygon.prevprev)
    
    s = (s1 + 2*s2 + 2*s3 + s4)/6.0

    p = inner_find_point p1, p2, p3, s2, s3, s, angle

    return p if @curr_polygon.has_point? p

    p "-------------- doh doh doh"
    return nil
    
  end

  def process_case3(angle)
    l1 = Edge.length(@curr_polygon.next, @curr_polygon.curr_point)
    l2 = Edge.length(@curr_polygon.curr_point, @curr_polygon.prev)
    
    p2 = @curr_polygon.curr_point
    p1 = @curr_polygon.prev
    a = l2
    
    index = @curr_polygon.curr_index
    
    if l1 < l2
      p1 = @curr_polygon.curr_point
      p2 = @curr_polygon.next
      a = l1
      
      index += 1
    end
    
    p_middle = Point.new((p1.x + p2.x)/2.0, (p1.y + p2.y)/2.0)
    
    #p = inner_find_point p_middle, p1, p2, a/2.0, a/2.0, Math.sqrt(3.0)*a/2.0, Math::PI
    
    m = 2.0*(p2.x - p1.x)
    n = 2.0*(p2.y - p1.y)
    k = (p2.x**2 + p2.y**2) - (p1.x**2 + p1.y**2)
    # k == m*x + n*y
    
    p = p2.y - p1.y
    q = p1.x - p2.x
    l = p1.x*p2.y - p2.x*p1.y + Math.sqrt(3.0)*a*a/2.0
    
    det = n*p - q*m
    x_new = (l*n - k*q)/det
    y_new = (k*p - l*m)/det
    
    p = Point.new(x_new, y_new)
    
    return [p, index] if @curr_polygon.has_point? p

    # find symmetric point
    p_new = Point.new(2.0*p.x - p_middle.x, 2.0*p.y - p_middle.y)
    return [p_new, index] if @curr_polygon.has_point? p_new
    
    p "doh doh doh 3"
    return nil
    
  end
  
  def inner_triangulate
    count = 0

    return @curr_polygon if @curr_polygon.size < 3
    
    if @curr_polygon.size == 3
      @triangles << @curr_polygon
      return @curr_polygon
    end

    if @curr_polygon.size == 4
      return process_polygon4(@curr_polygon)
    end

    # main loop
    until @curr_angles.empty?

      # @curr_polygon.gnuplot_draw
      
      # first stupid/simple situations
      if (@curr_polygon.size == 4)
        process_polygon4 @curr_polygon
        break
      end

      if (@curr_polygon.size == 3)
        @triangles << @curr_polygon
        save_triangle_points @curr_polygon, @triangles.size - 1
        break
      end

      break if (@curr_polygon.size < 3)

      # and now sophisticated stuff

      # at first delete element with smalest angle
      element = @curr_angles[0]
      point = element[0]
      angle = element[1]

      @curr_angles.delete_at 0
      @curr_polygon.curr_point = point

      # if it is just a cutoff - do it
      # and go to next loop iteration
      if angle < Math::PI/2.0
        #p "upper cutoff"
        cut_off_vertex(@curr_polygon.curr_index, angle)
        next
      end

      a1 = (Math::PI / 2.0)
      a2 = (5.0*Math::PI / 6.0)
      a3 = 2.0*Math::PI

      bad_point = false
      used_elements = []
      
      begin
        
        case angle
        when 0..a1 then
            #p "inner cutoff"
          bad_point = false
          cut_off_vertex(@curr_polygon.curr_index, angle)          
        when a1..a2 then
            #p 'case 2'
          p = process_case2 angle
          
          if p.nil?
            used_elements << [Point.new(point.x, point.y), angle]
            bad_point = true
          else
            bad_point = false

            # if all is ok - than we will insert
            # a new point - need to add first elements
            unless used_elements.empty?
              @curr_angles = used_elements + @curr_angles
              used_elements.clear
            end

            p1 = @curr_polygon.curr_point
            p2 = @curr_polygon.prev
            p3 = @curr_polygon.next

            tr1 = Polygon.copy([p1, p3, p])
            tr2 = Polygon.copy([p1, p, p2])
            
            @triangles << tr1
            save_triangle_points(tr1, @triangles.size - 1)

            @triangles << tr2
            save_triangle_points(tr2, @triangles.size - 1)

            angle1 = @curr_polygon.get_angle(@curr_polygon.curr_index - 1)
            angle2 = @curr_polygon.get_angle(@curr_polygon.curr_index + 1)

            @curr_polygon.replace_current!(p)

            # change angle of previous point
            new_angle = @curr_polygon.get_angle(@curr_polygon.curr_index - 1)
            change_alpha(@curr_polygon.prev, angle1, new_angle, @curr_angles)

            # add angle of next point
            new_angle = @curr_polygon.get_angle(@curr_polygon.curr_index + 1)
            change_alpha(@curr_polygon.next, angle2, new_angle, @curr_angles)

            # add angle of current point
            new_angle = @curr_polygon.get_angle(@curr_polygon.curr_index)
            insert_alpha_point new_angle, @curr_polygon.curr_point, @curr_angles          end
        when a2..a3 then
            return
            p 'case 3'
            arr = process_case3 angle
          if arr.nil?
            used_elements << [Point.new(point.x, point.y), angle]
            bad_point = true
          else
            bad_point = false            
            
            unless used_elements.empty?
              @curr_angles = used_elements + @curr_angles
              used_elements.clear
            end

            p = arr[0]
            index = arr[1]

            @curr_polygon.curr_index = index
            p2 = @curr_polygon.curr_point
            p1 = @curr_polygon.prev

            tr = Polygon.copy([p1, p2, p])
            @triangles << tr
            save_triangle_points(tr, @triangles.size - 1)

            @curr_polygon.insert_point index, p

            # insert angle of current vertex (a new vertex)
            new_angle = @curr_polygon.get_angle(index)
            insert_alpha_point new_angle, p, @curr_angles
            
            new_angle = @curr_polygon.get_angle(index - 1)
            change_alpha(@curr_polygon.prev, angle, new_angle, @curr_angles)
            
            new_angle = @curr_polygon.get_angle(index - 2)
            change_alpha(@curr_polygon.prevprev, angle, new_angle, @curr_angles)
          end
        end

        if (@curr_angles.size <= 4)
          break
        end
        # break if @curr_angles.size <= 4

        #p @curr_angles
        #p @curr_polygon

        if bad_point

          element = @curr_angles[0]
          point = element[0]
          angle = element[1]

          @curr_angles.delete_at 0

          @curr_polygon.curr_point = point

          p 'move next'
        else
          unless used_elements.empty?
            @curr_angles = used_elements + @curr_angles
            used_elements.clear
          end
        end
        
      end while bad_point
      
    end
    
  end
  
end
